### ISSUE TITLE
Cant properly include or exclude proto files

### ISSUE THREAD

**Issue by JustAnIssue** [google/protobuf-gradle-plugin#322]
I currently am not able to properly include or exclude certain directories containing proto files.
My current layout looks as follows:
```
Project
 - protobuf
   - dir1
   - dir2
   - dir3
 - subproject1
   - src
   - build.gradle
 - subproject2
   - src
   - build.gradle
```

In my root projects build.gradle I am only applying the protobuf plugin to subprojects that need it (by using `afterEvaluate {}`) like so:
```
afterEvaluate { Project project ->
        if` (project.hasProperty('proto') && project.property('proto')) {
                project.plugins.apply('com.google.protobuf')

                protobuf {
                    protoc {
                        artifact = 'com.google.protobuf:protoc:3.0.0'
                    }
                    generateProtoTasks.generatedFilesBaseDir = "$projectDir/gen-src"
                    generateProtoTasks {
                        all().each { task ->
                        }
                        ofSourceSet('main')
                    }
                }
                sourceSets {
                    main {
                        proto {
                            srcDir "$projectDir/../protobuf"
                        }
                    }
                }
        }
}
```
By doing `srcDir "$projectDir/../protobuf"` it is going to take all of the proto files in the folder. If I try doing an `include` or using `srcDir` multiple times it is not working properly, because the .proto file in protobuf/dir1 does `import "dir2/some.proto";`, which then has issues resolving the path.

**Comment by zhangkun83** [comment]
How does the `include` or multiple `srcDir` you have tried look like?

**Comment by JustAnIssue** [comment]
```
sourceSets {
    main {
        proto { // dir3 is still being used
            srcDir "$projectDir/../protobuf"
            include "$projectDir/../protobuf/dir1"
            include "$projectDir/../protobuf/dir2"
        }
    }
}
```

```
sourceSets {
    main {
        proto { // now i get an error because of 'import "dir2/some.proto";' in dir1/my.proto
            srcDir "$projectDir/../protobuf/dir1"
            srcDir "$projectDir/../protobuf/dir2"
        }
    }
}
```

Also tried using `include` twice but without any `srcDir` specified.
Is there a way to print what files protobuf is including right now? At the moment I am only able to `println sourceSets.main.proto.srcDirs`.

**Comment by voidzcy** [comment]
Can you try

```
sourceSets {
    main {
        proto {
            srcDir "$projectDir/../protobuf"
            exclude "$projectDir/../protobuf/dir3/"
        }
    }
}
```

**Comment by JustAnIssue** [comment]
Thanks for your answer, unfortunately this is not working for me.
`$projectDir` refers to 'subproject1' or 'subproject2', not to 'Project' (please note my structure at the very top), which is exactly how it is supposed to be. For that reason I do use `srcDir "$projectDir/../protobuf"` in order to step one level up.

By doing 
```
generateProtoTasks {
                        all().each { task ->
                            task.sourceFiles.each {
                                file -> println file
                            }
                        }
}
```
I now see that all my .proto files under "$projectDir/../protobuf/" are being used. I am now able to exclude a .proto file by doing `exclude "**/some.proto"`. However, switching to `include "**/reporting.proto"` is not doing anything, since it simply includes something that already has been included.
I would like to be able to specify a `srcDir` ($projectDir/../protobuf/) but not to include all .proto files (`**/*.proto`) automatically. Only those that i specify (`include "**/i_want_this_file_only_which_is_somewhere_in_protobuf.proto"`).

Basically I would like my code to look like:
```
sourceSets {
    main {
        proto { // only use '*.proto' files under 'dir1' or 'dir2'
            srcDir "$projectDir/../protobuf"
            include "**/dir1/*.proto"
            include "**/dir2/*.proto"
        }
    }
}
```
Which should ONLY take .proto files that are in a directory called 'dir1' or 'dir2', somewhere under 'protobuf'. How can I overwrite your `include **/*.proto` statement? 
Please note that excluding every directory besides 'dir1' and 'dir2' is working but not the approach i would like to use.

**Comment by voidzcy** [comment]
Ok. To do that, use [`setIncludes`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/util/PatternFilterable.html#setIncludes-java.lang.Iterable-) method (takes a list of String to override the existing include filter) instead of `include`.

**Comment by robvadai** [comment]
Same issue for me, I have a JAR dependency with literally only `*.proto` files in this structure:

```
|
|- configuration
|                  |----- config1.proto
|                  |----- config2.proto
|                  |--- modules
|                           |--------- module1.proto
|- modules
           |--------- module1.proto
```

```
sourceSets {
  main {
    proto {
      // In addition to the default 'src/main/proto'
      exclude "${buildDir}/**/main/modules"
      srcDir "${buildDir}/extracted-protos/main/configuration/modules"
      srcDir "${buildDir}/extracted-protos/main/configuration"
    }
  }
}
```

still fails with an error due to duplicated files, e.g.: `... Input is shadowed in the --proto_path ....` `Either use the latter file as your input or reorder the --proto_path so that the former file's location comes first.`

And no, I can't remove duplicated files, the source is external to our team.

Any help greatly appreciated.

**Comment by voidzcy** [comment]
You might want to try `exclude "${buildDir}/**/main/modules/"`, note the ending slash. [Gradle doc](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/util/PatternFilterable.html) says "Patterns ending with '/' or '\' will have '**' automatically appended", so I am suspecting the ending slash matters.
